SARS-CoV2 Wasterwater sequencing variant calling pipeline, based off Fontenele et al 2021

------------------------------------------------------------------------------------
protoco1:

step              	 tool               

1. trim       		    none                   

2. align       	      bwa                

3. trim primers    	  ivar trim         

4. call variants      (mpileup +) ivar variants   

5. build consensus   	(mpileup +) ivar consensus		 		

output:           	.tsv/.fa      
-------------------------------------------------------------------------------------
	
				

## Fontenele et al ##

```{bash}
#start w raw seqs

#align with bwa mem
#since every file gets the same RG line, fill line with file name .? (now needs double quotes for variable expansion)
(osu) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bwa mem -t 4 -M -R "@RG\tID:${bn}\tSM:${bn}\tLB:SWIFT\tPL:Illumina" ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}R1_001.fastq.gz ${bn}R2_001.fastq.gz > ${bn}.bwa.sam;
done

#Convert sam to bam
(osu) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sam; 
do bn=$(basename $i .bwa.sam);
samtools view -b ${bn}.bwa.sam -o ${bn}.bwa.bam
done

#switch to ivar conda env

#sort bam and make .bai index files for ivar trimming 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*R1_001.fastq.gz; 
do bn=$(basename $i R1_001.fastq.gz); 
samtools sort -o ${bn}.bwa.sorted.bam ${bn}.bwa.bam && samtools index ${bn}.bwa.sorted.bam; 
done

#ivar trim primers
#use default settings
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.bam;  
do bn=$(basename $i .bwa.sorted.bam); 
ivar trim -i ${bn}.bwa.sorted.bam -b ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed -p "${bn}.bwa.sorted.ivar.bam"; 
done

#sort again?
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.ivar.bam;  
do bn=$(basename $i .bwa.sorted.ivar.bam); 
samtools sort -o ${bn}.bwa.sorted.ivar.sorted.bam ${bn}.bwa.sorted.ivar.bam; 
done

#call variants:
#as per fontenele et al:
#no min cutoff frequency (-t 0 )
#min coverage (read depth) 20 (-m 20)
#default quality score (20)
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.ivar.sorted.bam;  
do bn=$(basename $i .bwa.sorted.ivar.sorted.bam); 
samtools mpileup -aa -A -d 0 -B -Q 20 --reference ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}.bwa.sorted.ivar.sorted.bam | ivar variants -t 0 -m 20 -p "${bn}.fontenele.variants" -r ~/SARS-COV2/sarscov2WuhanHu1.fasta -g ~/SARS-COV2/ivar/GCF_009858895.2_ASM985889v3_genomic.gff; 
done

#organize 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mkdir fontenele
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mv ** fontenele 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mv *variants* fontenele
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ cd fontenele

#generate consensus files 
#no min cutoff frequency (-t 0 )
#min coverage (read depth) 20 (-m 20)
#default quality sqore (20)
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1/fontenele$ for i in *.bwa.sorted.ivar.sorted.bam;  
do bn=$(basename $i .bwa.sorted.ivar.sorted.bam); 
samtools mpileup -aa -A -d 0 -Q 20 ${bn}.bwa.sorted.ivar.sorted.bam | ivar consensus -t 0 -m 20 -q 20 -p "${bn}.fontenele.consensus"; 
done

#rename consensus fastas
for i in *.fa;  
do bn=$(basename $i .fa);
sed 's/>/>fontenele_/g' ${bn}.fa > ${bn}.fasta;
done

```

```{bash}

## QC  ##

# genome coverage
for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam); 
samtools coverage ${bn}.sorted.bam > ${bn}.coverage.tsv; 
done

# average depth of covered bases
for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam); 
samtools depth ${bn}.sorted.bam | awk '{sum+=$3} END { print "Average = ",sum/NR}'; 
done

#depth of all dases
for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam); 
samtools depth ${bn}.sorted.bam > ${bn}.readdepth.tsv;
done

#to do:
num snvs
num snvs overlapping VOCs
```



#concatenate consensus files on local#
```{bash}

cmatt@DESKTOP-JN1E3J1 MINGW64 ~/Documents/sarscov2/ for i in *fontenele/consensus/*.fontenele.consensus.fasta;  
do bn=$(basename $i .fontenele.consensus.fasta);
cat ./fontenele/consensus/${bn}.fontenele.consensus.fasta ./critschristoph/consensus/${bn}.bowtie.sorted.sambamba.CC.fasta ./ivarmmp/consensus/${bn}.ivarmmp.consensus.fasta > ${bn}.consensus.fasta; 
done

```

