#goal: compare pipelines on environmental samples tested on both miseq and nextseq

------------------------------------------------------------------------------------
protocols:

step               OSU            (ivarXbwa      	Fontenele                 ivarXminimap2		       crits-christoff

trim               bbduk          bbduk      		none                      none				none

align              bwa            bwa       		bwa                       minimap2			bowtie2

trim primers      primer clip     ivar trim       	ivar trim                 ivar trim			none

call variants     gatk            ivar variants         ivar variants (mpileup)  ivar variants (mpileup)	inStrain

build consensus   ??		  ivar consensus 	ivar consensus		 ivar consensus			

output:           .vcf/fa         .tsv/.fa      	.tsv/.fa                 .tsv/.fa
-------------------------------------------------------------------------------------
	
				


```{bash}
#create directory structures for comaprison in follwing directiories: ivar, minimap, osu
#this will keep diretories the same level as the Miseq test and minimize edits
mkdir Environmental_Batch1_MiSeq
mkdir Environmental_Batch1_NextSeq 

#get data:

#MiSeq: 
cmatt5@crick:~/SARS-COV2/Batch1_MiSeq$ wget -r --level=10 -nH -nc --cut-dirs=3 --no-parent --reject "wget_index.html" --no-check-certificate --header "Cookie: sessionid=kis2uksxxdyksueredbr12mf5mxpq5hy;" https://bioshare.bioinformatics.ucdavis.edu/bioshare/wget/zp8306fews1o8ro/GSR_SWIFT/Environmental_Batch1_4-8-21/wget_index.html

#NextSeq:
cmatt5@crick:~/SARS-COV2/Batch1_NextSeq$ wget -r --level=10 -nH -nc --cut-dirs=3 --no-parent --reject "wget_index.html" --no-check-certificate --header "Cookie: sessionid=kis2uksxxdyksueredbr12mf5mxpq5hy;" https://bioshare.bioinformatics.ucdavis.edu/bioshare/wget/zp8306fews1o8ro/GSR_SWIFT/Environmental_Batch1_NextSeq/wget_index.html
#cut file prefix with for loop, example from Shannon in Brown lab:
for infile in reads/*R1_001.fastq.gz
do
bn=$(basename $infile R1_001.fastq.gz)
megahit -1 reads/${bn}_R1_001.fastq.gz -2 reads/${bn}_R2_001.fastq.gz -o megahit-min-count-3-assembly_${bn}/ -t 6 --min-count 3
done
```


## NextSeq ## 

## OSU pipeline ###

```{bash} 
#adapter trim with bbduk
(osu) cmatt5@c2-4:~/SARS-COV2/osu/Batch1_NextSeq$ for i in ../../Environmental_Batch1_NextSeq/*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bbduk.sh ktrim=r k=23 mink=11 hdist=1 qtrim=r1 trimq=15 minlength=30 maxns=0 t=4 in1=../../Environmental_Batch1_NextSeq/${bn}R1_001.fastq.gz in2=../../Environmental_Batch1_NextSeq/${bn}R2_001.fastq.gz out1=./${bn}R1_001.bbduk.fastq.gz out2=./${bn}R2_001.bbduk.fastq.gz ref=../../../bbmap/resources/adapters.fa;
done


for i *R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
awk '{if(NR%4==2) {count++; bases += length} } END{print bases/count}' ${bn}R1_001.fastq.gz; done

#already made bwa ref index. files are in ~/SARS-COV2

#keeps running out of space ('no space left on device')

#align with bwa mem
#since every file gets the same RG line, fill line with file name ..? (now needs double quotes for variable expansion)
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in ./*R1_001.bbduk.fastq.gz;
do bn=$(basename $i R1_001.bbduk.fastq.gz); 
bwa mem -t 4 -M -R "@RG\tID:${bn}\tSM:${bn}\tLB:SWIFT\tPL:Illumina" ../../sarscov2WuhanHu1.fasta ${bn}R1_001.bbduk.fastq.gz ${bn}R2_001.bbduk.fastq.gz > ${bn}.sam;
done

#sort sam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sam; 
do bn=$(basename $i .sam); 
samtools sort ${bn}.sam -o ${bn}.sorted.sam;
done

#clip primers with primerclip
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.sam; 
do bn=$(basename $i .sorted.sam); 
primerclip ../sarscov2_v1_masterfile.txt ${bn}.sorted.sam ${bn}.sorted.primerclipped.sam;
rm ${bn}.sorted.sam;
done

#Convert sam to bam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.primerclipped.sam; 
do bn=$(basename $i .sorted.primerclipped.sam);
samtools view -b ${bn}.sorted.primerclipped.sam -o ${bn}.bam;
done


#Sort reads
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.bam; 
do bn=$(basename $i .bam);
samtools sort ${bn}.bam -o ${bn}.sorted.bam;
done

#Index reads and get depth from BAM files
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
samtools index ${bn}.sorted.bam;
samtools depth -a ${bn}.sorted.bam | awk '$3 < 5' | cut -f 1,2 > output.depth
done

#call variants!
conda deactivate osu
conda activate gatk
(gatk) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
../gatk-4.2.0.0/gatk --java-options "-Xmx4g" HaplotypeCaller -R ../sarscov2WuhanHu1.fasta -ERC GVCF -stand-call-conf 20 --dont-use-soft-clipped-bases -mbq 20 --max-reads-per-alignment-start 0 --linked-de-bruijn-graph --recover-all-dangling-branches --sample-ploidy 1 --sample-name 'TEST' --read-filter 'NotDuplicateReadFilter' --input ${bn}.sorted.bam --output ${bn}.gvcf; 
done

#make consensus seqs
(gatk) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.vcf; 
do bn=$(basename $i .sorted.bam);
../gatk-4.2.0.0/gatk --java-options "-Xmx4g" FastaAlternateReferenceMaker -R ../sarscov2WuhanHu1.fasta -O ${bn}.osu.consensus.fasta -V "${bn}";
done

#try making consensus seqs with ivar consensus?
#flags translated from gatk 
# -mbq 20 == min base quality 20 -> -q 20
for i in ../*Environmental_Batch1/osuBAMs/*.sorted.bam;
do bn=$(basename $i .sorted.bam);

samtools mpileup -aa -A -d 0 -Q 0 ${bn}.bwa.sorted.ivar.sorted.bam | ivar consensus -t 0 -m 20 -q 20 -p "${bn}.fontenele.consensus"; 
done

```


## Fontenele et al (no trimming w bbduk) ##

```{bash}
#start w raw seqs

#align with bwa mem
#since every file gets the same RG line, fill line with file name .? (now needs double quotes for variable expansion)
(osu) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bwa mem -t 4 -M -R "@RG\tID:${bn}\tSM:${bn}\tLB:SWIFT\tPL:Illumina" ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}R1_001.fastq.gz ${bn}R2_001.fastq.gz > ${bn}.bwa.sam;
done

#Convert sam to bam
(osu) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sam; 
do bn=$(basename $i .bwa.sam);
samtools view -b ${bn}.bwa.sam -o ${bn}.bwa.bam
done

#switch to ivar conda env

#sort bam and make .bai index files for ivar trimming 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*R1_001.fastq.gz; 
do bn=$(basename $i R1_001.fastq.gz); 
samtools sort -o ${bn}.bwa.sorted.bam ${bn}.bwa.bam && samtools index ${bn}.bwa.sorted.bam; 
done

#ivar trim primers
#use default settings
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.bam;  
do bn=$(basename $i .bwa.sorted.bam); 
ivar trim -i ${bn}.bwa.sorted.bam -b ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed -p "${bn}.bwa.sorted.ivar.bam"; 
done

#sort again?
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.ivar.bam;  
do bn=$(basename $i .bwa.sorted.ivar.bam); 
samtools sort -o ${bn}.bwa.sorted.ivar.sorted.bam ${bn}.bwa.sorted.ivar.bam; 
done

#call variants:
#as per fontenele et al:
#no min cutoff frequency (-t 0 )
#min coverage (read depth) 20 (-m 20)
#default quality score (20)
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bwa.sorted.ivar.sorted.bam;  
do bn=$(basename $i .bwa.sorted.ivar.sorted.bam); 
samtools mpileup -aa -A -d 0 -B -Q 20 --reference ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}.bwa.sorted.ivar.sorted.bam | ivar variants -t 0 -m 20 -p "${bn}.fontenele.variants" -r ~/SARS-COV2/sarscov2WuhanHu1.fasta -g ~/SARS-COV2/ivar/GCF_009858895.2_ASM985889v3_genomic.gff; 
done

#organize 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mkdir fontenele
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mv ** fontenele 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ mv *variants* fontenele
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ cd fontenele

#generate consensus files 
#no min cutoff frequency (-t 0 )
#min coverage (read depth) 20 (-m 20)
#default quality sqore (20)
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1/fontenele$ for i in *.bwa.sorted.ivar.sorted.bam;  
do bn=$(basename $i .bwa.sorted.ivar.sorted.bam); 
samtools mpileup -aa -A -d 0 -Q 20 ${bn}.bwa.sorted.ivar.sorted.bam | ivar consensus -t 0 -m 20 -q 20 -p "${bn}.fontenele.consensus"; 
done

#rename consensus fastas
for i in *.fa;  
do bn=$(basename $i .fa);
sed 's/>/>fontenele_/g' ${bn}.fa > ${bn}.fasta;
done

```




## ivarXminimap2 ##

#as per https://github.com/CDCgov/SARS-CoV-2_Sequencing/tree/master/protocols/BFX-UT_ARTIC_Illumina

```{bash}
conda activate osu

#do not trim raw reads

# line as from artic github (piping not working for some reason):
minimap2 -ax sr -t 4 -o ${bn}.mmp.sam ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}R1_001.fastq.gz ${bn}R2_001.fastq.gz | samtools sort | samtools view -F 4 -o ${bn}.mmp.sorted.bam;
done

#align with minimap2 
#code from Utah DoH https://github.com/CDCgov/SARS-CoV-2_Sequencing/tree/master/protocols/BFX-UT_ARTIC_Illumina
# minimap2 -ax sr -t {threads} -o aligned/{sample}.sam artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta {input.read1} {input.read2} | samtools sort | samtools view -F 4 -o {sample}.sorted.bam
#since every file gets the same RG line, fill line with file name ..? (now needs double quotes for variable expansion)
(covid) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
minimap2 -ax sr -t 4 -o ${bn}.mmp.sam ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}R1_001.fastq.gz ${bn}R2_001.fastq.gz;
done

#sort sam 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in ./*.mmp.sam
do bn=$(basename $i .mmp.sam); 
samtools sort -o ${bn}.mmp.sorted.sam ${bn}.mmp.sam;
done

#sam to bam
for i in *.mmp.sam; 
do bn=$(basename $i .mmp.sam);
samtools view -F 4 -b ${bn}.mmp.sorted.sam -o ${bn}.mmp.sorted.bam
done

#ivar trim primers
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.mmp.sorted.bam;  
do bn=$(basename $i .mmp.sorted.bam); ivar trim -i ${bn}.mmp.sorted.bam -b ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed -p "${bn}.mmp.sorted.ivar.bam"; 
done

#sort again?
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.mmp.sorted.ivar.bam;  
do bn=$(basename $i .mmp.sorted.ivar.bam); samtools sort -o ${bn}.mmp.sorted.ivar.sorted.bam ${bn}.mmp.sorted.ivar.bam; 
done

#call variants:
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.mmp.sorted.ivar.sorted.bam;  
do bn=$(basename $i .mmp.sorted.ivar.sorted.bam); 
samtools mpileup  -A -d 1000 -B -Q 0 --reference ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}.mmp.sorted.ivar.sorted.bam | ivar variants -p "${bn}.ivar.mmp.variants" -r ~/SARS-COV2/sarscov2WuhanHu1.fasta -g ~/SARS-COV2/ivar/GCF_009858895.2_ASM985889v3_genomic.gff; 
done

#generate consensus files 
(ivar) cmatt5@c2-18:/group/diazmunozlab/share/NextSeq_Environmental_Batch1/fontenele$ for i in *.mmp.sorted.ivar.sorted.bam;  
do bn=$(basename $i .mmp.sorted.ivar.sorted.bam); 
samtools mpileup  -A -d 1000 -B -Q 0  ${bn}.mmp.sorted.ivar.sorted.bam | ivar consensus -p "${bn}.ivarmmp.consensus"; 
done

#rename consensus fastas
for i in *.fa;  
do bn=$(basename $i .fa);
sed 's/>/>mmpXivar_/g' ${bn}.fa > ${bn}.fasta;
done

```

## crits-cristoff ##

```{bash}
#make sambam sript
(christ) cmatt5@c2-3:~/SARS-COV2$ vim sambam
#copy/paste from github (https://github.com/alexcritschristoph/wastewater_sarscov2)
`
#!/bin/sh
#v0.1: Simple wrapper to convert sam files to sorted bam files - Rohan Sachdeva
samtools view -bu | samtools sort -@6 -l 6 -
`

#make executable?
(christ) cmatt5@c2-3:~/SARS-COV2$ chmod +x sambam
(christ) cmatt5@c2-3:/group/diazmunozlab/share/NextSeq_Environmental_Batch1 $mkdir critschristoff

#make bowtie2 index for ref fasta
~/SARS-COV@$ bowtie2-build ~/SARS-COV2/sarscov2WuhanHu1.fasta sarscov2WuhanHu1

#no trim 
#sambam doesnt wor, do sam->bam and sort separately

#align w bowtie2
(christ) cmatt5@c2-3:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bowtie2 -p 20 -x sarscov2WuhanHu1 -1 ${bn}R1_001.fastq.gz -2 ${bn}R2_001.fastq.gz -S ${bn}.bowtie.sam;
done

#sam to bam
(christ) cmatt5@c2-3:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bowtie.sam; 
do bn=$(basename $i .bowtie.sam);
samtools view -b ${bn}.bowtie.sam -o ${bn}.bowtie.bam
done


(christ) cmatt5@c2-3:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bowtie.bam;  
do bn=$(basename $i .bowtie.bam); 
samtools sort -o ${bn}.bowtie.sorted.bam ${bn}.bowtie.bam -@6 -l 6; 
done

(christ) cmatt5@c2-3:/group/diazmunozlab/share/NextSeq_Environmental_Batch1$ for i in *.bowtie.sorted.bam;  
do bn=$(basename $i .bowtie.sorted.bam); 
sambamba markdup -r ${bn}.bowtie.sorted.bam ${bn}.bowtie.sorted.sambamba.bam; 
done


for i in *.bowtie.sorted.sambamba.bam;  
do bn=$(basename $i .bowtie.sorted.sambamba.bam); 
inStrain profile ${bn}.bowtie.sorted.sambamba.bam -l 0.9 ../sarscov2WuhanHu1.fasta -o ${bn}.bowtie.sorted.sambamba.bam.instrain;
done

#get-consensus
for i in *.bowtie.sorted.sambamba.bam;  
do bn=$(basename $i .bowtie.sorted.sambamba.bam); 
python ~/SARS-COV2/get_consensus.py ./${bn}.bowtie.sorted.sambamba.bam ~/SARS-COV2/sarscov2WuhanHu1.fasta;
done 

for i in *.fasta;  
do bn=$(basename $i .fasta);
sed 's/>/>crits-christoph_/g' ${bn}.fasta > $CC.{bn}.fasta;
done

```

## QC ?? ##

```{bash}

#average depth of covered bases
for i in *.bwa.sorted.ivar.sorted.bam; do bn=$(basename $i .bwa.sorted.ivar.sorted.bam); samtools depth ${bn}.bwa.sorted.ivar.sorted.bam |  awk '{sum+=$3} END { print "Average = ",sum/NR}'; done

NextSeq		MiSeq
632.46		90.9051
1106.7		173.152
1245.93		178.622
1655.21		280.695
1667.21		273.818
788.496		109.768
1678.13		270.527
1137.95		186.635
714.669		93.3433
1915.14		278.379
1581.58		219.86
1525.25		309.669
2770.25		506.866
1703.76		271.324
1512.85		225.141
647.034		87.6829
2727.74		499.554
2377.33		410.155
1815.75		254.279
2210.21		490.293
1774.19		272.706
974.363		136.245
890.714		127.131
520.666		68.9691
```

## get NextFlow NF-Core Viral Recon pipeline working ##
```{bash}
~/nextflow run nf-core/viralrecon --help

 ~/nextflow run nf-core/viralrecon --input ./sample_test.csv --platform illumina --protocol amplicon --genome NC_045512.2 --fasta ~/SARS-COV2/sarscov2WuhanHu1.fasta --skip_kraken2 --skip_fastp --skip_cutadapt --callers ivar --min_mapped_reads 20 --filter_duplicates --skip_assembly --primer_bed ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed --max_memory 40.GB --primer_left_suffix F --primer_right_suffix R --ivar_trim_offset 5

primer_set SWIFT --primer_bed ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed 
```



#concatenate consensus files on local#
```{bash}

cmatt@DESKTOP-JN1E3J1 MINGW64 ~/Documents/sarscov2/ for i in *fontenele/consensus/*.fontenele.consensus.fasta;  
do bn=$(basename $i .fontenele.consensus.fasta);
cat ./fontenele/consensus/${bn}.fontenele.consensus.fasta ./critschristoph/consensus/${bn}.bowtie.sorted.sambamba.CC.fasta ./ivarmmp/consensus/${bn}.ivarmmp.consensus.fasta > ${bn}.consensus.fasta; 
done

````
