#goal: compare pipelines on environmental samples tested on both miseq and nextseq

```{bash}
#create directory structures for comaprison in follwing directiories: ivar, minimap, osu
#this will keep diretories the same level as the Miseq test and minimize edits
mkdir Environmental_Batch1_MiSeq
mkdir Environmental_Batch1_NextSeq 

#get data:
#MiSeq: 
cmatt5@crick:~/SARS-COV2/Batch1_MiSeq$ wget -r --level=10 -nH -nc --cut-dirs=3 --no-parent --reject "wget_index.html" --no-check-certificate --header "Cookie: sessionid=kis2uksxxdyksueredbr12mf5mxpq5hy;" https://bioshare.bioinformatics.ucdavis.edu/bioshare/wget/zp8306fews1o8ro/GSR_SWIFT/Environmental_Batch1_4-8-21/wget_index.html
#NextSeq:
cmatt5@crick:~/SARS-COV2/Batch1_NextSeq$ wget -r --level=10 -nH -nc --cut-dirs=3 --no-parent --reject "wget_index.html" --no-check-certificate --header "Cookie: sessionid=kis2uksxxdyksueredbr12mf5mxpq5hy;" https://bioshare.bioinformatics.ucdavis.edu/bioshare/wget/zp8306fews1o8ro/GSR_SWIFT/Environmental_Batch1_NextSeq/wget_index.html

#cut file prefix with for loop, example from Shannon in Brown lab:

for infile in reads/*R1_001.fastq.gz
do
bn=$(basename $infile R1_001.fastq.gz)
megahit -1 reads/${bn}_R1_001.fastq.gz -2 reads/${bn}_R2_001.fastq.gz -o megahit-min-count-3-assembly_${bn}/ -t 6 --min-count 3
done





# **OSU PIPELINE**

#use tmux:
tmux new -s var
srun -J gatk -t 18:00:00 --mem=25gb -c 1 --pty bash

conda activate osu

## NEXTSEQ ##

#adapter trim with bbduk
(osu) cmatt5@c2-4:~/SARS-COV2/osu/Batch1_NextSeq$ for i in ../../Environmental_Batch1_NextSeq/*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bbduk.sh ktrim=r k=23 mink=11 hdist=1 qtrim=r1 trimq=15 minlength=30 maxns=0 t=4 in1=../../Environmental_Batch1_NextSeq/${bn}R1_001.fastq.gz in2=../../Environmental_Batch1_NextSeq/${bn}R2_001.fastq.gz out1=./${bn}R1_001.bbduk.fastq.gz out2=./${bn}R2_001.bbduk.fastq.gz ref=../../../bbmap/resources/adapters.fa;
done

#already made bwa ref index. files are in ~/SARS-COV2

#keeps running out of space ('no space left on devide'), remvove files as i go

#align with bwa mem
#since every file gets the same RG line, fill line with file name ..? (now needs double quotes for variable expansion)
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in ./*R1_001.bbduk.fastq.gz;
do bn=$(basename $i R1_001.bbduk.fastq.gz); 
bwa mem -t 4 -M -R "@RG\tID:${bn}\tSM:${bn}\tLB:SWIFT\tPL:Illumina" ../../sarscov2WuhanHu1.fasta ${bn}R1_001.bbduk.fastq.gz ${bn}R2_001.bbduk.fastq.gz > ${bn}.sam;
rm ${bn}R1_001.bbduk.fastq.gz;
rm ${bn}R2_001.bbduk.fastq.gz
done

#sort sam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sam; 
do bn=$(basename $i .sam); 
samtools sort ${bn}.sam -o ${bn}.sorted.sam;
rm ${bn}.sam;
done

#clip primers with primerclip
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.sam; 
do bn=$(basename $i .sorted.sam); 
primerclip ../sarscov2_v1_masterfile.txt ${bn}.sorted.sam ${bn}.sorted.primerclipped.sam;
rm ${bn}.sorted.sam;
done

#Convert sam to bam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.primerclipped.sam; 
do bn=$(basename $i .sorted.primerclipped.sam);
samtools view -b ${bn}.sorted.primerclipped.sam -o ${bn}.bam;
rm ${bn}.sorted.primerclipped.sam;
done


#Sort reads
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.bam; 
do bn=$(basename $i .bam);
samtools sort ${bn}.bam -o ${bn}.sorted.bam;
rm ${bn}.sam;
done

#Index reads and get depth from BAM files
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
samtools index ${bn}.sorted.bam;
samtools depth -a ${bn}.sorted.bam | awk '$3 < 5' | cut -f 1,2 > output.depth
done

#call variants!
conda deactivate osu
conda activate gatk
(gatk) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
../gatk-4.2.0.0/gatk --java-options "-Xmx4g" HaplotypeCaller -R ../sarscov2WuhanHu1.fasta -ERC GVCF -stand-call-conf 20 --dont-use-soft-clipped-bases -mbq 20 --max-reads-per-alignment-start 0 --linked-de-bruijn-graph --recover-all-dangling-branches --sample-ploidy 1 --sample-name 'TEST' --read-filter 'NotDuplicateReadFilter' --input ${bn}.sorted.bam --output ${bn}.gvcf; 
done

****************************************************************************




## MISEQ ## 
#adapter trim with bbduk
(osu) cmatt5@c2-4:~/SARS-COV2/osu/Batch1_MiSeq$ for i in ../../Environmental_Batch1_MiSeq/*R1_001.fastq.gz;
do bn=$(basename $i R1_001.fastq.gz); 
bbduk.sh ktrim=r k=23 mink=11 hdist=1 qtrim=r1 trimq=15 minlength=30 maxns=0 t=4 in1=../../Environmental_Batch1_MiSeq/${bn}R1_001.fastq.gz in2=../../Environmental_Batch1_MiSeq/${bn}R2_001.fastq.gz out1=./${bn}R1_001.bbduk.fastq.gz out2=./${bn}R2_001.bbduk.fastq.gz ref=../../../bbmap/resources/adapters.fa;
done

#align with bwa mem
#since every file gets the same RG line, fill line with file name ..? (now needs double quotes for variable expansion)
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in ./*R1_001.bbduk.fastq.gz;
do bn=$(basename $i R1_001.bbduk.fastq.gz); 
bwa mem -t 4 -M -R "@RG\tID:${bn}\tSM:${bn}\tLB:SWIFT\tPL:Illumina" ../../sarscov2WuhanHu1.fasta ${bn}R1_001.bbduk.fastq.gz ${bn}R2_001.bbduk.fastq.gz > ${bn}.sam;
rm ${bn}R1_001.bbduk.fastq.gz;
rm ${bn}R2_001.bbduk.fastq.gz
done

#sort sam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in *.sam; 
do bn=$(basename $i .sam); 
samtools sort ${bn}.sam -o ${bn}.sorted.sam;
rm ${bn}.sam;
done

#clip primers with primerclip
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in *.sorted.sam; 
do bn=$(basename $i .sorted.sam); 
primerclip ../sarscov2_v1_masterfile.txt ${bn}.sorted.sam ${bn}.sorted.primerclipped.sam;
rm ${bn}.sorted.sam;
done

#Convert sam to bam
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in *.sorted.primerclipped.sam; 
do bn=$(basename $i .sorted.primerclipped.sam);
samtools view -b ${bn}.sorted.primerclipped.sam -o ${bn}.bam;
rm ${bn}.sorted.primerclipped.sam;
done

#Sort reads
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in *.bam; 
do bn=$(basename $i .bam);
samtools sort ${bn}.bam -o ${bn}.sorted.bam;
rm ${bn}.bam;
done

#Index reads and get depth from BAM files
(osu) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_MiSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
samtools index ${bn}.sorted.bam;
samtools depth -a ${bn}.sorted.bam | awk '$3 < 5' | cut -f 1,2 > output.depth
done

#call variants!
conda deactivate osu
conda activate gatk
(gatk) cmatt5@c2-18:~/SARS-COV2/osu/Batch1_NextSeq$ for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
../gatk-4.2.0.0/gatk --java-options "-Xmx4g" HaplotypeCaller -R ../sarscov2WuhanHu1.fasta -ERC GVCF -stand-call-conf 20 --dont-use-soft-clipped-bases -mbq 20 --max-reads-per-alignment-start 0 --linked-de-bruijn-graph --recover-all-dangling-branches --sample-ploidy 1 --sample-name 'TEST' --read-filter 'NotDuplicateReadFilter' --input ${bn}.sorted.bam --output ${bn}.gvcf; 
done



```






#ivar

```{bash}

#OSU BAMS: 

#trim primers: 
for i in *.sorted.bam; 
do bn=$(basename $i .sorted.bam);
ivar trim -i ${bn}.sorted.bam -b ~/SARS-COV2/ivar/NC_045512.2.v2.primers.bed -p "${bn}.ivartrimmed.bam";
done

#need to sort trimmed files?
for i in *.ivartrimmed.bam; 
do bn=$(basename $i .ivartrimmed.bam);
samtools sort -o ${bn}.ivartrimmed.sorted.bam ${bn}.ivartrimmed.bam;
done

#run variant calling:
for i in *.ivartrimmed.sorted.bam; 
do bn=$(basename $i .ivartrimmed.sorted.bam);
samtools mpileup -aa -A -d 0 -B -Q 0 --reference ~/SARS-COV2/sarscov2WuhanHu1.fasta ${bn}.ivartrimmed.sorted.bam | ivar variants -p "${bn}.variants" -r ~/SARS-COV2/sarscov2WuhanHu1.fasta -g ~/SARS-COV2/ivar/GCF_009858895.2_ASM985889v3_genomic.gff;
done




```
